fastplotlib/layouts/_record_mixin.py

```diff
@@ -0,0 +1,238 @@
+from typing import *
+from pathlib import Path
+from multiprocessing import Queue, Process
+from time import time
+
+try:
+    import av
+except ImportError:
+    HAS_AV = False
+else:
+    HAS_AV = True
+
+
+class VideoWriterAV(Process):
+    """Video writer, uses PyAV in an external process to write frames to disk"""
+    def __init__(
+            self,
+            path: Union[Path, str],
+            queue: Queue,
+            fps: int,
+            width: int,
+            height: int,
+            codec: str,
+            pixel_format: str,
+            options: dict = None
+    ):
+        super().__init__()
+        self.queue = queue
+
+        self.container = av.open(path, mode="w")
+
+        self.stream = self.container.add_stream(codec, rate=fps, options=options)
+
+        # in case libx264, trim last rows and/or column
+        # because libx264 doesn't like non-even number width or height
+        if width % 2 != 0:
+            width -= 1
+        if height % 2 != 0:
+            height -= 1
+
+        self.stream.width = width
+        self.stream.height = height
+
+        self.stream.pix_fmt = pixel_format
+
+    def run(self):
+        while True:
+            if self.queue.empty():  # no frame to write
+                continue
+
+            frame = self.queue.get()
+
+            # recording has ended
+            if frame is None:
+                self.container.close()
+                break
+
+            frame = av.VideoFrame.from_ndarray(
+                frame[:self.stream.height, :self.stream.width],  # trim if necessary because of x264
+                format="rgb24"
+            )
+
+            for packet in self.stream.encode(frame):
```

Comment:
what are these lines doing? what is a packet?

Reply:
also `.mux()` function?

Reply:
I don't know exactly, I copied from the pyav docs example for writing video to file. It works so.